_drafts/practical_zephyr_04_dt_practice.md

```diff
@@ -0,0 +1,1384 @@
+
+We've learned that UART is used for the console output and thus for `printk`, so let's use some `printk` calls to make use of it. We'll output `"tick"` each time the LED is turned on, and `"tock"` each time it is turned off:
+
+```c
+void main(void)
+{
+    int err   = 0;
+    bool tick = true;
+
+    if (!gpio_is_ready_dt(&led))
+    {
+        printk("Error: LED pin is not available.\n");
+        return;
+    }
+
+    err = gpio_pin_configure_dt(&led, GPIO_OUTPUT_ACTIVE);
+    if (err != 0)
+    {
+        printk("Error %d: failed to configure LED pin.\n", err);
+        return;
+    }
+
+    while (1)
+    {
+        (void) gpio_pin_toggle_dt(&led);
+        k_msleep(SLEEP_TIME_MS);
+
+        if (tick != false) { printk("Tick\n"); }
+        else { printk("Tock\n"); }
+        tick = !tick;
+    }
+}
+```
+
+> **Note:** If you've been following along, make sure to set the UART node's status back to `"okay"` in case you've disabled it in the last section.
+
+
+### Peripheral pin control in Zephyr
+
+To see how pins are assigned to our UART peripheral, we need to look into our board's DTS file. There, we find `pinctrl-<x>` properties for the `&uart0` node:
+
+`zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts`
+
+```dts
+&uart0 {
+  compatible = "nordic,nrf-uarte";
+  status = "okay";
+  current-speed = <115200>;
+  pinctrl-0 = <&uart0_default>;
+  pinctrl-1 = <&uart0_sleep>;
+  pinctrl-names = "default", "sleep";
+};
+```
+
+Before we look at how `pinctrl` works, you might ask yourself why we need another mechanism when we have `gpios`, and what the difference between the two is. Couldn't we just use `gpios` and, e.g., pass _phandles_ for all the pins we need for our UART interface?
+
+For some MCUs, like Nordic's nRF series, this might work since pin multiplexing for such MCUs is not restricted: On Nordic's MCUs it is possible to assign any functionality to any pin. Other MCUs (e.g., STM32) typically restrict this and clearly define possible alternate pin functions for all pins.
+
+MCU manufacturers can be quite creative when it comes to pin multiplexing, and before Zephyr 3.0 pin multiplexing was entirely vendor-specific. With Zephyr 3.0 (and mostly adopted with version 3.1) Zephyr adopted the [Linux `pinctl`](https://www.kernel.org/doc/html/v4.13/driver-api/pinctl.html) concept as a _standardized_ way of assigning peripheral functions to pins in Devicetree - and called it `pinctrl`.
+
+There is, of course, still an overlap between `gpios` and `pinctrl` since both associate nodes with pins and their parameters (e.g., pull resistors). As a rule of thumb, `pinctrl` is generally used if the pin goes to a **peripheral** in the `/soc` node, whereas `gpios` are used if the pin is used by the application.
```

Comment:
```suggestion
There is, of course, still an overlap between `gpios` and `pinctrl` since both associate nodes with pins and their parameters (e.g., pull resistors). As a rule of thumb, `pinctrl` is generally used if the pin goes to a **peripheral** in the `/soc` node, whereas `gpios` are used if the pin is used directly by the application.
```